File: javascript/ch1/Ex123.js
// Load with: var ExXXX = require('./ExXXX')
// Run with: ExXXX.xxx()

let sq = n => n * n;

let divides = (a, b) => (b % a) == 0;

let find_divisor = (n, test_divisor) => {
    if (sq(test_divisor) > n) {
        return n;
    } else if (divides(test_divisor, n)) {
        return test_divisor;
    } else {
        return find_divisor(n, next(test_divisor));
    }
};

let next = n => n==2 ? 3 : n+2;

let smallestDiv = n => find_divisor(n, 2);

exports.smallest_divisor = smallestDiv;

exports.test_it = function() {
    console.log(smallestDiv(199));
    console.log(smallestDiv(1999));
    console.log(smallestDiv(19999));
};

File: java/src/ch1/Ex123.java
package ch1;

public class Ex123 {
    public static void main(String[] args) {
        System.out.println(new SmallestDivisor(199).value());
        System.out.println(new SmallestDivisor(1999).value());
        System.out.println(new SmallestDivisor(19999).value());
        System.out.println("done");
    }

    public static int square(int n) {
        return n * n;
    }

    public static boolean divides(int a, int b) {
        return 0 == b % a;
    }

    public static int next(int n) {
        return a == 2 ? 3 : n+2;
    }

    static class SmallestDivisor extends CachedValue<Integer> {
        int n;
        SmallestDivisor(int number) {
            n = number;
        }

        @Override
        public Value<Integer> compute() {
            val = findDivisor(n, 2);
            return this;
        }

        int findDivisor(int n, int test_divisor) {
            if (square(test_divisor) > n) {
                return n;
            } else if (divides(test_divisor, n)) {
                return test_divisor;
            } else {
                return findDivisor(n, next(test_divisor));
            }
        }
    }
}
